smarts/sstudio/genhistories.py

```diff
@@ -450,6 +460,159 @@ def column_val_in_row(self, row, col_name):
+            self._log.error(errmsg)
+            raise ValueError(errmsg)
+
+        for i in range(len(scenario.tracks)):
+            vehicle_id = scenario.tracks[i].id
+            vehicle_type = self._lookup_agent_type(scenario.tracks[i].object_type)
+            rows = []
+
+            # First pass -- extract data
+            for j in range(len(scenario.timestamps_seconds)):
+                obj_state = scenario.tracks[i].states[j]
+                vel = np.array([obj_state.velocity_x, obj_state.velocity_y])
+
+                row = {}
+                row["valid"] = obj_state.valid
+                row["vehicle_id"] = vehicle_id
+                row["type"] = vehicle_type
+                row["length"] = obj_state.length
+                row["height"] = obj_state.height
+                row["width"] = obj_state.width
+                row["sim_time"] = scenario.timestamps_seconds[j]
+                row["position_x"] = obj_state.center_x
+                row["position_y"] = obj_state.center_y
+                row["heading_rad"] = obj_state.heading - math.pi / 2
+                row["speed"] = np.linalg.norm(vel)
+                row["lane_id"] = 0
+                row["is_ego_vehicle"] = 1 if i == scenario.sdc_track_index else 0
+                rows.append(row)
+
+            # Second pass -- align timesteps to 10 Hz and interpolate trajectory data if needed
+            for j in range(len(scenario.timestamps_seconds)):
+                row = rows[j]
+                if row["valid"] == False:
+                    continue
+                timestep = 0.1
+                time_current = row["sim_time"]
+                time_expected = round(j * timestep, 3)
+                time_error = time_current - time_expected
+
+                if time_error == 0:
+                    continue
+
+                # Align to expected timestep
+                row["sim_time"] = time_expected
+
+                if time_error > 0:
+                    # If first element, we can't interpolate backwards
+                    if j == 0:
+                        continue
+
+                    # Interpolate backwards using previous timestep
+                    prev_row = rows[j - 1]
+                    if prev_row["valid"]:
+                        prev_time = prev_row["sim_time"]
+                        t = (time_expected - prev_time) / (time_current - prev_time)
+                        row["speed"] = lerp(prev_row["speed"], row["speed"], t)
+                        row["position_x"] = lerp(
+                            prev_row["position_x"], row["position_x"], t
+                        )
+                        row["position_y"] = lerp(
+                            prev_row["position_y"], row["position_y"], t
+                        )
+                        row["heading_rad"] = lerp(
+                            prev_row["heading_rad"], row["heading_rad"], t
+                        )
```

Comment:
We might need to shift this back to be between 0 and 2pi (or -pi and pi?).

Reply:
`row["heading_rad"] = obj_state.heading - math.pi / 2`.
They are shifted back first and then interpolated.

Reply:
But I think the interpolation can still cause values to be outside of this range.  For example, if the value at time `t` is `1.5*pi` and the value at time `t+0.08` is `2*pi`, then the interpolated value for time `t+0.1` will be bigger than `2*pi`.